From 433c6fbd8b9eaaa5a94684775bbece20a5d7c7e2 Mon Sep 17 00:00:00 2001
From: Dejan Muhamedagic <dejan@hello-penguin.com>
Date: Wed, 10 Sep 2014 17:33:46 +0200
Subject: [PATCH] Medium: clplumbing: don't abort on bad input

Receiving garbage from the peer shouldn't cause us to abort.
We'll just disconnect the other peer. Thanks to Lars Ellenberg
for finding the flaw and suggesting the improvement.

Origin: upstream, https://github.com/ClusterLabs/cluster-glue/commit/433c6fbd8b9eaaa5a94684775bbece20a5d7c7e2
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1895355
Reviewed-by: Rafael David Tinoco <rafaeldtinoco@ubuntu.com>
Last-Update: 2020-09-15
---
 lib/clplumbing/ipcsocket.c | 11 +++++++++--
 lib/clplumbing/ocf_ipc.c   |  6 +++---
 2 files changed, 12 insertions(+), 5 deletions(-)

diff --git a/lib/clplumbing/ipcsocket.c b/lib/clplumbing/ipcsocket.c
index 679b70d5..14c35043 100644
--- a/lib/clplumbing/ipcsocket.c
+++ b/lib/clplumbing/ipcsocket.c
@@ -1388,8 +1388,15 @@ socket_resume_io_read(struct IPC_CHANNEL *ch, int* nbytes, gboolean read1anyway)
 			*nbytes = msg_len;
 			nmsgs = ipc_bufpool_update(pool, ch, msg_len, ch->recv_queue) ;
 
-			SocketIPCStats.ninqueued += nmsgs;
-
+			if (nmsgs < 0) {
+				/* we didn't like the other side */
+				cl_log(LOG_ERR, "socket_resume_io_read: "
+					   "disconnecting the other side");
+				ch->ch_status = IPC_DISCONNECT;
+				retcode = IPC_FAIL;
+			} else {
+				SocketIPCStats.ninqueued += nmsgs;
+			}
 		}
 	}
 
diff --git a/lib/clplumbing/ocf_ipc.c b/lib/clplumbing/ocf_ipc.c
index 28c168cc..c243934c 100644
--- a/lib/clplumbing/ocf_ipc.c
+++ b/lib/clplumbing/ocf_ipc.c
@@ -445,8 +445,8 @@ ipc_bufpool_update(struct ipc_bufpool* pool,
 		if (head->magic != HEADMAGIC) {
 			GList* last = g_list_last(rqueue->queue);
 			cl_log(LOG_ERR, "ipc_bufpool_update: "
-			       "magic number in head does not match."
-			       "Something very bad happened, abort now, farside pid =%d",
+			       "magic number in head does not match. "
+			       "Something very bad happened, farside pid =%d",
 			       ch->farside_pid);
 			cl_log(LOG_ERR, "magic=%x, expected value=%x", head->magic, HEADMAGIC);
 			ipc_bufpool_display(pool);
@@ -456,7 +456,7 @@ ipc_bufpool_update(struct ipc_bufpool* pool,
 				IPC_Message* m = (IPC_Message*)last;
 				ipcmsg_display(m);
 			}
-			abort();
+			return -1;
 		}
 
 		if ( head->msg_len > MAXMSG) {
-- 
2.27.0

